File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_general.py
from collections import deque

def input_graph():
    n = int(input("Nhap so dinh: "))
    m = int(input("Nhap so canh: "))
    directed = input("Do thi co huong? (y/n): ").strip().lower() == 'y'
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input("Nhap canh u v: ").split())
        adj[u].append(v)
        if not directed:
            adj[v].append(u)
    return n, adj

def bfs_general(start, adj, n):
    visited = [False] * n
    queue = deque([start])
    visited[start] = True
    print("Thu tu duyet:", end=" ")
    while queue:
        u = queue.popleft()
        print(u, end=" ")
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                queue.append(v)

if __name__ == "__main__":
    n, adj = input_graph()
    s = int(input("Nhap dinh bat dau: "))
    bfs_general(s, adj, n)

File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_general.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
using namespace std;

void input_graph(int& n, vector<vector<int>>& adj, bool& directed) {
    int m;
    cout << "Nhap so dinh: ";
    cin >> n;
    cout << "Nhap so canh: ";
    cin >> m;
    cout << "Do thi co huong? (y/n): ";
    string ans; cin >> ans;
    directed = (ans == "y" || ans == "Y");
    adj.resize(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cout << "Nhap canh u v: ";
        cin >> u >> v;
        adj[u].push_back(v);
        if (!directed) adj[v].push_back(u);
    }
}

void bfs_general(int start, const vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    cout << "Thu tu duyet: ";
    while (!q.empty()) {
        int u = q.front(); q.pop();
        cout << u << " ";
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

int main() {
    int n;
    bool directed;
    vector<vector<int>> adj;
    input_graph(n, adj, directed);
    int s;
    cout << "Nhap dinh bat dau: ";
    cin >> s;
    bfs_general(s, adj, n);
    return 0;
}